packages/evolution-backend/src/api/survey.user.routes.ts

```diff
@@ -99,6 +100,11 @@ export default (authorizationMiddleware, loggingMiddleware: InterviewLoggingMidd
                 const ip = (req as any).clientIp;
                 const timestamp = moment().unix();
 
+                // Get the current session data to send to client, then set to undefined again
+                const session = req.session;
+                const sessionClientValues = session.clientValues?.valuesByPath;
+                delete session.clientValues;
```

Comment:
Wondering, is there possible race condition here if we process 2 requests at the same time ?

Reply:
Sure, there are many possible race conditions:

* 2 requests could trigger the long callback, which will both run together at the same time. Any of them could finish first and update the client value
* Here we set and delete the sessionClientValue at the beginning of the request. And it will only be saved at the end of it (after a few async call). In theory, the client should wait patiently for the request to return before doing anything else, but a quick participant could refresh the page, which would trigger some updates, which could cause another update of the values and maybe a second set of values will come back to the client, or this one. But the most common case would be that 2 requests would return the same data, which wouldn't be dramatic.
* Also, about the comment if the mechanism can be abused, yes indeed it can, since it's documented to be for long-lasting operation, a malefic participant could log in and trigger the conditions that cause this operation to be run and if the programmer of the callback does not plan for some cancelation/throttling mechanism, it could indeed cause denial of service on the server. But that is the case for any serverFieldUpdate callback with longer running operations, it's just that since the deferred callback returns to the client faster, the UI becomes responsive again faster and can be "abused" faster too (though one could achieve the same effect with many calls to curls or something in a script, so the responsive of the UI is not a condition), and is more sensible to bad programming/bad flow of survey (if the callback condition is met too frequently).

The most important is that whatever is saved in the survey responses matches what will be returned to the client. The race condition is not as much here (because the current happy path survey flow prevents 2 requests from the same participant at the same time) than during the data saving phase (the deferredSave in the interview.ts file, or in the deferredUpdateCallback below) if 2 or more long lasting operations fill the same data at the same time.

With the current, it's the survey programmer's responsibility to take care that the long running operations do not race because we don't know that a long lasting operation is running, we don't know that a deferred update callback will eventually be called.

We could instead have the server field update callback register the long lasting operation to a path, so that evolution could cancel the previous operation if another one for the same path occurs. That will reduce, though not totally eliminate the risk of race.

Reply:
And now that I pressed send on the previous comment, I think another more race-mitigating approach would be that instead of saving in the session the full data to send to the client, we could save only the paths that were modified, and just before sending the responses back, we get the data for those paths directly from the recently loaded/saved interview. That would ensure we send the data that is currently in the interview.

In parallel, in the deferredUpdateCallback call below, we should reload the session and save it in a db transaction, to make sure we have the latest data and here too it should be saved right away, after deleting the clientValues to avoid overwriting the deferred callback data.

Or we implement mutexes of response paths!